bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/ProjectEncodingPreferenceChangeListener.java

```diff
@@ -0,0 +1,196 @@
+/*******************************************************************************
+ * Copyright (c) 2022 IBM Corporation and others.
+ *
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Ingo Mohr - Issue #166 - Add Preference to Turn Off Warning-Check for Project Specific Encoding
+ *******************************************************************************/
+package org.eclipse.core.internal.resources;
+
+import java.util.*;
+import org.eclipse.core.internal.utils.Messages;
+import org.eclipse.core.resources.*;
+import org.eclipse.core.runtime.*;
+import org.eclipse.core.runtime.jobs.Job;
+import org.eclipse.core.runtime.preferences.IEclipsePreferences;
+import org.eclipse.core.runtime.preferences.IEclipsePreferences.IPreferenceChangeListener;
+import org.eclipse.core.runtime.preferences.IEclipsePreferences.PreferenceChangeEvent;
+import org.eclipse.core.runtime.preferences.IScopeContext;
+import org.osgi.service.component.annotations.*;
+
+@Component(service = {})
+public class ProjectEncodingPreferenceChangeListener implements IPreferenceChangeListener {
+
+	private static final String MARKER_TYPE = ValidateProjectEncoding.MARKER_TYPE;
+
+	@Reference
+	private IWorkspace workspace;
+
+	@Reference
+	private ILog log;
+
+	@Reference(target = IScopeContext.BUNDLE_SCOPE_FILTER)
+	IScopeContext bundleScope;
+
+	private IEclipsePreferences eclipsePreferences;
+
+	@Activate
+	public void register() {
+		eclipsePreferences = bundleScope.getNode(""); //$NON-NLS-1$
+		eclipsePreferences.addPreferenceChangeListener(this);
+	}
+
+	@Deactivate
+	public void unregister() {
+		eclipsePreferences.removePreferenceChangeListener(this);
+	}
+
+	@Override
+	public void preferenceChange(PreferenceChangeEvent event) {
+		if (ResourcesPlugin.PREF_MISSING_ENCODING_MARKER_SEVERITY.equals(event.getKey())) {
+			final int newSeverity = event.getNewValue() != null ? Integer.parseInt((String) event.getNewValue())
+					: PreferenceInitializer.PREF_MISSING_ENCODING_MARKER_SEVERITY_DEFAULT;
+			final int oldSeverity = event.getOldValue() != null ? Integer.parseInt((String) event.getOldValue())
+					: PreferenceInitializer.PREF_MISSING_ENCODING_MARKER_SEVERITY_DEFAULT;
+
+			if (isIgnore(newSeverity)) {
```

Comment:
The if/else with lot of custom code below in this class could be simply replaced to 
```
if (newSeverity != oldSeverity) {
	ValidateProjectEncoding.scheduleWorkspaceValidation((Workspace) workspace);
}
```

And then, looking on what is left (just the prefs listener), the code should be simply moved to `CharsetManager` which is started automatically anyway, so no extra component needed, and it can register this tiny listener in `org.eclipse.core.internal.resources.CharsetManager.startup(IProgressMonitor)` and the remaining code in this method above will do the job.

Reply:
> tiny listener in org.eclipse.core.internal.resources.CharsetManager.startup(IProgressMonitor)

What about loose coupling via OSGi service and single responsibility (as in SOLID)? Adding the pref listener to the ``ChartSetManager``  would again hardwire that listener to another class - which (as I understood from a former review comment) is not what we want.

Reply:
> > tiny listener in org.eclipse.core.internal.resources.CharsetManager.startup(IProgressMonitor)
> 
> What about loose coupling via OSGi service and single responsibility (as in SOLID)? Adding the pref listener to the `ChartSetManager` would again hardwire that listener to another class - which (as I understood from a former review comment) is not what we want.

Sounds good, but fully missing the point. 

The `ChartSetManager` **is** about managing that stuff for which the preference is added. Why should it violate any rules if exact that manager not only triggers creation of markers but also updates them? With few extra lines of code vs extra "loose coupled" OSGI service that adds a lot more code, complexity and overhead for startup to do exact same thing. That makes no sense.

Reply:
> Sounds good, but fully missing the point.
>
> The ChartSetManager is about managing that stuff for which the preference is added. Why should it violate any rules if exact that manager not only triggers creation of markers but also updates them? With few extra lines of code vs extra "loose coupled" OSGI service that adds a lot more code, complexity and overhead for startup to do exact same thing. That makes no sense.

@iloveeclipse , @laeubi , @merks 
Ok. I'd say you decide on this together and I'll make the change. I'm not going to make it w/o the consent of at least 2 of you as I already refactored how to connect the pref change listener before and I don't want to invest my free time in refactoring the same code back and forth.

Reply:
> I already refactored how to connect the pref change listener before and I don't want to invest my free time in refactoring the same code back and forth.

There will be no back and forth. There is literally no need in any OSGI service or component here.
There is no need in big refactoring either. All what is needed now to throw away **all** of complex and unnecessary code wiring that extra component which only purpose is to activate and add a small listener and moving that small listener into the CharsetManager where it belongs to.

"If the only tool you have is a hammer, you tend to see every problem as a nail.” 
Here is no nails for OSGI, the code change to add a listener is trivial and doesn't require OSGI.

Reply:
@iloveeclipse This code is neither complex nor a "hammer" it is separation of concerns and all those "WhateverMangers" that are tightly integrated and mangled into the Workspace init with different concerns are better separated instead of "just adding a few lines" her and there and calling back and forth, carefully tweaking order of init calls, keeping care of cleanup manually and so on.

So why not take the opportunity and start right now? Why these reservations and discomfort? All this stuff was added to finally solve the dreaded early Resource-Plugin-Activation problem that was not solved for > 10 years so why not use it and recommend its usage to give examples for other adopters where it is maybe not just a few lines?

@ingomohr I did not requested any changes but was asked how to do it with services and I gave the necessary hints how to do it without static access, but that's not a requirement (from the project) and just to show people now ways instead of "we have always done it that way".

So everyone is free to ignore my suggestions here (that's why I left a comment versus a real review) but then please don't complain that there are "gazillion" of references to the static stuff, or that workspace init sometimes break, is slow or wrecked in any other way :-)

Reply:
Thank you for the response, @iloveeclipse  and @laeubi.

I'm completely w/ @laeubi on this matter. Still, I refactored the code to drop the service as requested by @iloveeclipse and supported by @merks - simply because we should put an end to this PR. I'm just a contributor and can only decide for my own plug-ins. For those, my take-away is that I can declare those listeners via OSGi to gain all the benefits @laeubi listed. Thank you for that! :)